huaweicloud/services/iotda/resource_huaweicloud_iotda_product.go

```diff
@@ -0,0 +1,604 @@
+
+	if err != nil {
+		return diag.Errorf("error updating IoTDA product: %s", err)
+	}
+
+	return ResourceProductRead(ctx, d, meta)
+}
+
+func ResourceProductDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	deleteOpts := &model.DeleteProductRequest{ProductId: d.Id()}
+	_, err = client.DeleteProduct(deleteOpts)
+	if err != nil {
+		return diag.Errorf("error deleting IoTDA product: %s", err)
+	}
+
+	return nil
+}
+
+func buildProductCreateParams(d *schema.ResourceData) *model.CreateProductRequest {
+	req := model.CreateProductRequest{
+		Body: &model.AddProduct{
+			ProductId:           utils.StringIgnoreEmpty(d.Get("product_id").(string)),
+			Name:                d.Get("name").(string),
+			DeviceType:          d.Get("device_type").(string),
+			ProtocolType:        d.Get("protocol").(string),
+			DataFormat:          d.Get("data_type").(string),
+			ManufacturerName:    utils.StringIgnoreEmpty(d.Get("manufacturer_name").(string)),
+			Industry:            utils.StringIgnoreEmpty(d.Get("industry").(string)),
+			Description:         utils.StringIgnoreEmpty(d.Get("description").(string)),
+			AppId:               utils.StringIgnoreEmpty(d.Get("space").(string)),
+			ServiceCapabilities: *buildServices(d.Get("services").([]interface{})),
+		},
+	}
+	return &req
+}
+
+func buildProductUpdateParams(d *schema.ResourceData) *model.UpdateProductRequest {
+	req := model.UpdateProductRequest{
+		ProductId: d.Id(),
+		Body: &model.UpdateProduct{
+			Name:                utils.String(d.Get("name").(string)),
+			DeviceType:          utils.String(d.Get("device_type").(string)),
+			ProtocolType:        utils.String(d.Get("protocol").(string)),
+			DataFormat:          utils.String(d.Get("data_type").(string)),
+			ManufacturerName:    utils.String(d.Get("manufacturer_name").(string)),
+			Industry:            utils.String(d.Get("industry").(string)),
+			Description:         utils.String(d.Get("description").(string)),
+			AppId:               utils.String(d.Get("space").(string)),
+			ServiceCapabilities: buildServices(d.Get("services").([]interface{})),
+		},
+	}
+	return &req
+}
+
+func buildServices(raw []interface{}) *[]model.ServiceCapability {
+	rst := make([]model.ServiceCapability, len(raw))
+	for i, v := range raw {
+		service := v.(map[string]interface{})
+		rst[i] = model.ServiceCapability{
+			ServiceId:   service["id"].(string),
+			ServiceType: service["type"].(string),
+			Description: utils.StringIgnoreEmpty(service["description"].(string)),
+			Properties:  buildServiceProperties(service["properties"]),
+			Commands:    buildServiceCommands(service["commands"]),
+		}
+	}
+
+	return &rst
+}
+
+func buildServiceCommands(commandsRaw interface{}) *[]model.ServiceCommand {
+	if commandsRaw == nil {
```

Comment:
commandsRaw will not be nil, should check length

Reply:
ok